File: 4_Structured_GE/struct_ge.py
# Computational Linear Algebra #4 Structured Gaussian Elimination
# By: Nick Space Cowboy

import numpy as np

class Cowboy_Lin_Alg(object):
	def solve_utri(self, Utri, b):
		n = len(Utri) # row dimension of the Utri matrix
		x = np.zeros_like(b, dtype=np.float64)
		for i in range(n - 1, -1, -1):	# loop to iterate through row index
			x[i] += b[i] / Utri[i,i]
			for j in range(n-1, i, -1):	# loop to iterate through the off diagonal Sum part
				x[i] += (- (Utri[i, j] * x[j])) / Utri[i,i]
		return x
		
	def SGE(self, Am, b):
		n = len(Am)
		l = np.zeros([n, n], dtype=np.float64)
		for i in range(0, n, 1):
			for j in range(i+1, n, 1):
				l[j,i] = A[j,i] / A[i,i]
				A[j] = A[j] - (l[j,i] * A[i])
				b[j] = b[j] - (l[j,i] * b[i])
		return A, b
			
if __name__ == "__main__":		
	A = np.array(np.random.randint (0,100,(4,4)), dtype=np.float64)
	Ac = A.copy()
	print("A = ")
	print(A)
	b = np.array(np.random.randint(0, 100, (4,1)), dtype=np.float64)	
	print("b = ")
	print(b)
	cla = Cowboy_Lin_Alg()
	cla.SGE(A,b)
	x = cla.solve_utri(A, b)
	print("U = ")
	print(A)
	print("c = ")
	print(b)
	print("x = ")
	print(x)
	print("Check Ax = ")
	print(Ac.dot(x))

File: 4_Structured_GE/struct_ge.m
# Computational Linear Algebra Ep#4 Structured Gaussian Elimination
# By: Nick Space Cowboy

clear all;
close all;

function x = solve_utri(A,b) # creates the function solve_utri
	n = length(b);	# To help with indexing
	x = zeros(n,1);	# creates a x vector to be filled
	for k=n:-1:2	# starts iterating through matrix 
		x(k) = b(k)/A(k,k);
		b(1:k-1) = b(1:k-1) - x(k)*A(1:k-1,k);
	end
	x(1) = b(1)/A(1,1);
endfunction

function [A,b] = sge(A,b) 
	# takes an input A matrix and b vector of a linear system
	# and converts the system to Ux=c through structured gaussian elimination
	n = length(b);
	L = zeros(n,n);
	for k=1:1:n
		for l=(k+1):1:n
			L(l,k) = A(l,k) / A(k,k); # get and store the multipliers
			A(l,:) = A(l,:) - (L(l,k) * A(k, :));# use the multipliers to zero out a position of A
			b(l) = b(l) - (L(l,k) * b(k)); # use the multipliers to change a value of b
		end
	end
endfunction


b = randi(100, 5, 1) # generates random b vector
A = randi(100, 5, 5)# genertates random A matrix
[U,c] = sge(A,b)# converts the system to Ux=c through SGE
x = solve_utri(U,c)	# solves the system with our created function

check = A*x	# checks that our result is in fact correct
